### DISCUSSION TITLE
[`reportUninitializedInstanceVariable`] `dataclass` with `ABC`: Variables defined in abstract base class are not initialized in final class

### DISCUSSION THREAD

**Discussion by mark-henry01** [microsoft/pyright#8202] [Q&A] [answered]
It would be nice to allow usage of `ABC`/abstract `dataclasses` like in following snippet:

```python
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import final, override


@dataclass
class Parent(ABC):
    @abstractmethod
    def foo() -> str: ...

    p1: str


@final
@dataclass
class Child(Parent): # currently errors
    p2: int

    @override
    def foo():
        return "bar"
```

Currently `Child` reports an error:

> Variables defined in abstract base class are not initialized in final class "Child"
>   Instance variable "p1" is defined in abstract base class "Parent" but not initialized [reportUninitializedInstanceVariable]

Does anything speak against allowing above constellation with `reportUninitializedInstanceVariable`? Note I did not use `Protocol` for nominal typing support. Thanks!

(I am not sure of the specification regarding interplay between `ABC` and `dataclasses`, hence creating discussion topic rather than an issue.)

**Comment by erictraut** [comment]
I think this it's reasonable to add special cases for `dataclass`, `dataclass_transform`, and `NamedTuple`. These all involve synthesized `__init__` methods. 

I'll create a new enhancement request from this discussion thread.

**Reply by erictraut** [reply, marked as answer]
Here's the new [enhancement request](https://github.com/microsoft/pyright/issues/8205).

**Reply by mark-henry01** [reply]
Much appreciated @erictraut !
